### PR TITLE
adding nf-core/validation

### PR DESCRIPTION
**Pull request by louperelo**
Modified the necessary files to add nf-core/validation with the help of @mirpedrol 

## PR checklist

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/funcscan/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, also make a PR on the nf-core/funcscan _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [ ] Make sure your code lints (`nf-core lint`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

**Comment by mirpedrol** [comment]
Some of the errors are due to the meta map in the input channel bein an ImmutableMap. This means that is a map which can’t be modified in place, because this can introduce some bugs due to nextflow process being run in parallel. This can be deactivated and go back to a normal and modifiable map, but I think it’s a good opportunity to update the code to avoid possible future problems with these map operations.

For example, [this](https://github.com/nf-core/funcscan/actions/runs/5391241169/jobs/9830380858?pr=285) error points that you can't `.put()` items into an ImmutableMap.
In your code, [line 152](https://github.com/nf-core/funcscan/blob/master/workflows/funcscan.nf#L152C78-L152C78), you are adding a new value to the meta map `meta['longest_contig'] = Integer.parseInt(length)`, which uses the method `.put()`. One option to allow the modification of this is map is adding the new value with the operator `+` which is allowed in ImmutableMaps (this operator constructs a new Map based on the original one, so it is not modifying the original map).
See an example [here](https://github.com/nf-core/crisprseq/blob/master/workflows/crisprseq.nf#L103). I think the code in this case should look something like 
```
def new_meta = meta + ['longest_contig': Integer.parseInt(length)]
```

**Comment by jfy133** [comment]
> Some of the errors are due to the meta map in the input channel bein an ImmutableMap. This means that is a map which can’t be modified in place, because this can introduce some bugs due to nextflow process being run in parallel. This can be deactivated and go back to a normal and modifiable map, but I think it’s a good opportunity to update the code to avoid possible future problems with these map operations.
> 
> For example, [this](https://github.com/nf-core/funcscan/actions/runs/5391241169/jobs/9830380858?pr=285) error points that you can't `.put()` items into an ImmutableMap. In your code, [line 152](https://github.com/nf-core/funcscan/blob/master/workflows/funcscan.nf#L152C78-L152C78), you are adding a new value to the meta map `meta['longest_contig'] = Integer.parseInt(length)`, which uses the method `.put()`. One option to allow the modification of this is map is adding the new value with the operator `+` which is allowed in ImmutableMaps (this operator constructs a new Map based on the original one, so it is not modifying the original map). See an example [here](https://github.com/nf-core/crisprseq/blob/master/workflows/crisprseq.nf#L103). I think the code in this case should look something like
> 
> ```
> def new_meta = meta + ['longest_contig': Integer.parseInt(length)]
> ```

OMg this is awesome! Ever since that was brought up and Rob's talk I've been super paranoid about this! It's really cool nf-validation will help us find these places where we could be borking things :D

**Comment by mirpedrol** [comment]
It is probable that we have to unable the ImmutableMaps though, the new class has some errors when using `-resume` :( See [discussion](https://github.com/nextflow-io/nf-validation/issues/52). But at least we can already fix this line here :)

**Comment by jfy133** [comment]
Tested:

1. Duplicate sample IDs
2. Duplicate FASTAs
3. Spaces in name
4. Duplicate lines
5. File does not exist
6. Added support for `fna.gz`

**Comment by jfy133** [comment]
@louperelo could you and @mirpedrol look again, if all good we can merge into `dev`, and then I will start work again on the ORF input

File: CHANGELOG.md [hunk 1]
```diff
@@ -8,6 +8,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### `Added`
 
 - [#290](https://github.com/nf-core/funcscan/pull/290) Merged pipeline template of nf-core/tools version 2.9, updated references. (by @jfy133)
+- [#285](https://github.com/nf-core/funcscan/pull/285) Added support for `fna.gz` input FASTA files (by @louperelo, @mirpedrol, @jfy133)
```
Comment (mirpedrol) [inline CHANGELOG.md:11]: Is this description correct?

File: workflows/funcscan.nf [hunk 2]
```diff
@@ -3,6 +3,7 @@
     PRINT PARAMS SUMMARY
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 */
+include { validateParameters; paramsHelp; paramsSummaryLog; paramsSummaryMap; fromSamplesheet } from 'plugin/nf-validation'
 
 include { paramsSummaryLog; paramsSummaryMap } from 'plugin/nf-validation'
```
Comment (mirpedrol) [inline workflows/funcscan.nf:9]: include statements are duplicated here, I think it's enough with including `paramsSummaryLog`, `paramsSummaryMap` and `fromSamplesheet`

**Review by mirpedrol — APPROVED** [review]
LGTM :)

File: workflows/funcscan.nf [hunk 3]
```diff
@@ -3,6 +3,7 @@
...
```
Comment (jfy133) [inline workflows/funcscan.nf:8]:
```suggestion
include { paramsSummaryLog; paramsSummaryMap; fromSamplesheet } from 'plugin/nf-validation'
```

File: workflows/funcscan.nf (hunk 2 above)
Comment (jfy133) [reply workflows/funcscan.nf:9]: not the help?

Comment (mirpedrol) [reply workflows/funcscan.nf]: `paramsHelp` and `validateParameters` used in `main.nf` since the template update

File: CHANGELOG.md (hunk 1 above)
Comment (mirpedrol) [reply CHANGELOG.md:11]: Maybe "create input channel with Channel.fromSamplesheet"? Or similar. 
But this one sounds good too, I didn't see you changed the pattern and was confused :)

File: assets/schema_input.json [hunk 4]
```diff
@@ -10,27 +10,19 @@
...
-                    },
-                    {
-                        "type": "string",
-                        "maxLength": 0
-                    }
-                ]
+                "format": "file-path",
+                "exists": true,
+                "pattern": "^\\S+\\.f(ast|n)?a\\.gz$",
+                "errorMessage": "Fasta file for reads 1 must be provided, cannot contain spaces and must have extension '.fa.gz', '.fna.gz' or '.fasta.gz'",
```
Comment (louperelo) [inline assets/schema_input.json:22]: We do not have reads 1 and 2, right?
```suggestion
                "errorMessage": "Fasta file for reads must be provided, cannot contain spaces and must have extension '.fa.gz', '.fna.gz' or '.fasta.gz'",
```

**Review by louperelo — COMMENTED** [review]
Just found a tiny typo. Otherwise all good.

File: CHANGELOG.md (hunk 1 above)
Comment (jfy133) [reply CHANGELOG.md:11]:
```suggestion
- [#285](https://github.com/nf-core/funcscan/pull/285) Use nf-validation for samplesheet checking and added support for `fna.gz` input FASTA files (by @louperelo, @mirpedrol, @jfy133)
```

**Pull request merged**